text/0006-shipyard-next-generation.md

```diff
@@ -0,0 +1,317 @@
+
+```yaml
+rollback:
+  triggers:
+  - hardening.deployment.finished:
+      selector:
+        matchLabels:
+          status: failed
+  - hardening.release.finished:
+      selector:
+        matchExpressions:
+          - {key: status, operator: In, values: [warning, failed]}
+  tasks:
+  - name: rollback
+```
+
+### Shipyard controller
+
+This KEP also proposes the implementation of a Keptn core component - the so-called *shipyard controller*. This controller implements the functionality described above. Besides, this controller is responsible for keeping sequences atomic. This means that a sequence is executed based on the shipyard, which was configured when the sequence was triggered. In other words, an update of a shipyard (and its containing sequences) does NOT impact ongoing sequences.
+
+### Keptn CloudEvents
+
+Based on the propsed concepts, changes on the Keptn CloudEvents are necessary.
+
+**(1) Reserved space in CloudEvent payload:**
+
+Before the shipyard controller sends out a `[task].triggered` event for a specific task, the controller adds a reserved space to the event payload. This reserved space is named after the task and can be consumed by the unit (Keptn-service) that executes the task. For example, a *deployment* task gets the reserved space of *deployment* in the payload (i.e., data field) of the CloudEvent. Per default, this reserved space is empty. Expect for tasks that have properties defined in the shipyard (e.g., `strategy: blue_green`) - for those tasks the properties are added to the reserved space:
+  
+```
+{
+  "type": "sh.keptn.event.deployment.triggered",
+  ...
+  "data": {
+    "deployment": {
+      "strategy": "blue_green",
+      "traffic": "100",
+    }
+  }
+}
+```
+
+*Implementation details:* 
+* Keptn-services can add any information to the reserved space. For example:
+  ```
+  {
+    "type": "sh.keptn.event.deployment.triggered",
+    ...
+    "data": {
+      "deployment": {
+        "strategy": "blue_green",
+        "traffic": "100",
+        "deploymentURI":"https://my-service.domain.com/" // Property added by Keptn-service
+      }
+    }
+  }
+  ```
+
+* The event selectors for Keptn-services, as proposed in [First version of a Keptn uniform](https://github.com/keptn/enhancement-proposals/pull/7), work based on the information provided in this reserved space. For example:
```

Comment:
please add *KEP 07:* to the uniform link

Reply:
I added it and changed the link text.